edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent/MetricsWorker.cs

```diff
@@ -0,0 +1,129 @@
+
+    public class MetricsWorker : IDisposable
+    {
+        readonly IMetricsScraper scraper;
+        readonly IMetricsStorage storage;
+        readonly IMetricsUpload uploader;
+        readonly ISystemTime systemTime;
+        readonly AsyncLock scrapeUploadLock = new AsyncLock();
+        static readonly ILogger Log = Logger.Factory.CreateLogger<MetricsScraper>();
+
+        DateTime lastUploadTime = DateTime.MinValue;
+
+        // This acts as a local buffer. It stores the previous value of every metric.
+        // If the new value for that metric is unchanged, it doesn't write the duplicate value to disk.
+        Dictionary<int, Metric> metrics = new Dictionary<int, Metric>();
+
+        PeriodicTask scrape;
+        PeriodicTask upload;
+
+        public MetricsWorker(IMetricsScraper scraper, IMetricsStorage storage, IMetricsUpload uploader, ISystemTime systemTime = null)
+        {
+            this.scraper = scraper;
+            this.storage = storage;
+            this.uploader = uploader;
+            this.systemTime = systemTime ?? SystemTime.Instance;
+        }
+
+        public void Start(TimeSpan scrapingInterval, TimeSpan uploadInterval)
+        {
+            this.scrape = new PeriodicTask(this.Scrape, scrapingInterval, scrapingInterval, Log, "Scrape");
+            this.upload = new PeriodicTask(this.Upload, uploadInterval, uploadInterval, Log, "Scrape");
+        }
+
+        async Task Scrape(CancellationToken cancellationToken)
+        {
+            using (await this.scrapeUploadLock.LockAsync(cancellationToken))
+            {
+                Log.LogInformation("Scraping Metrics");
+                List<Metric> metricsToPersist = new List<Metric>();
+
+                foreach (var moduleResult in await this.scraper.ScrapeAsync(cancellationToken))
+                {
+                    IEnumerable<Metric> scrapedMetrics = PrometheousMetricsParser.ParseMetrics(this.systemTime.UtcNow, moduleResult.Value);
+
+                    foreach (Metric scrapedMetric in scrapedMetrics)
+                    {
+                        // Get the previous scrape for this metric
+                        if (this.metrics.TryGetValue(scrapedMetric.HashNameAndTag(), out Metric oldMetric))
+                        {
+                            // If the metric is unchanged, do nothing
+                            if (oldMetric.Value.Equals(scrapedMetric.Value))
+                            {
+                                continue;
+                            }
+
+                            // if the metric has changed, write the previous metric to disk
+                            metricsToPersist.Add(oldMetric);
+                        }
+
+                        // if new metric or metric changed, save to local buffer
+                        this.metrics[scrapedMetric.HashNameAndTag()] = scrapedMetric;
+                    }
+                }
+
+                Log.LogInformation("Scraped Metrics");
```

Comment:
nit: maybe report how many entries in scrapedMetric as well.

Reply:
this.metrics stores last metrics value of a specific metrics key; and only persist it when next time comes in.  Will it be lost if there is any crash?